### PR TITLE
fix: Update to use non-zero `reserveFactor` (SC-404)

### PR DESCRIPTION
**Pull request by lucas-manuel**

### CODE REVIEW

File: test/FlashLoan.t.sol [hunk 1]
```diff
@@ -855,7 +882,7 @@ contract FlashLoanSuccessTests is FlashLoanTestBase {
             currentVariableBorrowRate: borrowRate,
             currentStableBorrowRate:   0,
             lastUpdateTimestamp:       1,
-            accruedToTreasury:         0,
+            accruedToTreasury:         0,
```
Comment (barrutko) [inline test/FlashLoan.t.sol:885]:
```suggestion
            accruedToTreasury:         0,
```

File: test/FlashLoanSimple.t.sol [hunk 2]
```diff
@@ -396,33 +397,48 @@ contract FlashLoanSimpleSuccessTests is FlashLoanSimpleTestBase {
...
-        assertLt(liquidityRate, 0.00525e27);
+        assertLt(liquidityRate, 0.0049875e27);
         assertEq(borrowRate,    0.052476418612348581178374884e27);
-        assertEq(liquidityRate, 0.005198143190440621251948266e27);
+        assertEq(liquidityRate, 0.004938236030918590189350852e27);
 
         poolParams.variableBorrowIndex       = expectedBorrowIndex;
-        poolParams.currentLiquidityRate      = liquidityRate + 1;  // Rounding
-        poolParams.currentVariableBorrowRate = borrowRate + 1;  // Rounding
+        poolParams.currentLiquidityRate      = liquidityRate + 2;  // Rounding
```
Comment (barrutko) [inline test/FlashLoanSimple.t.sol:414]:
```suggestion
        poolParams.currentLiquidityRate      = liquidityRate + 2;  // Rounding
```

File: test/FlashLoanSimple.t.sol [hunk 3]
```diff
@@ -448,47 +464,55 @@ contract FlashLoanSimpleSuccessTests is FlashLoanSimpleTestBase {
...
-        assertEq(liquidityRate, 0.005198143190440621251948266e27);
+        assertEq(liquidityRate, 0.004938236030918590189350852e27);
+
+        // 100 + full 10% premium (0.5% protocol fee accounted with accruedToTreasury)
+        assetParams.aTokenBalance = 910 ether;
 
         poolParams.variableBorrowIndex       = expectedBorrowIndex;
-        poolParams.currentLiquidityRate      = liquidityRate + 1;  // Rounding
+        poolParams.currentLiquidityRate      = liquidityRate + 2;  // Rounding
         poolParams.currentVariableBorrowRate = borrowRate + 1;  // Rounding
```
Comment (barrutko) [inline test/FlashLoanSimple.t.sol:484]:
```suggestion
        poolParams.currentLiquidityRate      = liquidityRate + 2;  // Rounding
        poolParams.currentVariableBorrowRate = borrowRate + 1;     // Rounding
```

File: test/RepayWithATokens.t.sol [hunk 4]
```diff
@@ -315,9 +315,9 @@ contract RepayWithATokensConcreteTests is RepayWithATokensTestBase {
         AssertPoolReserveStateParams memory poolParams = AssertPoolReserveStateParams({
             asset:                     address(borrowAsset),
             liquidityIndex:            1e27,
-            currentLiquidityRate:      0.03125e27,  // Fully utilized
+            currentLiquidityRate:      0.0296875e27,  // 3.125% * 95%
```
Comment (barrutko) [inline test/RepayWithATokens.t.sol:318]:
```suggestion
            currentLiquidityRate:      0.0296875e27,  // 3.125% * 95%
```

File: test/RepayWithATokens.t.sol [hunk 5]
```diff
@@ -366,9 +366,9 @@ contract RepayWithATokensConcreteTests is RepayWithATokensTestBase {
...
```
Comment (barrutko) [inline test/RepayWithATokens.t.sol:369]: There might be more of that
```suggestion
            currentLiquidityRate:      0.0296875e27,  // 3.125% * 95%
```

File: test/RepayWithATokens.t.sol [hunk 6]
```diff
@@ -660,24 +662,28 @@ contract RepayWithATokensConcreteTests is RepayWithATokensTestBase {
...
         ( borrowRate, liquidityRate ) = _getUpdatedRates(1, remainingSupply);
 
+        liquidityRate -= 2;  // Rounding
+        borrowRate    -= 1;  // Rounding
+
         // Diff from 500 is small enough that it rounds to clean numbers here (unlike repay tests where full utilization caused more debt)
-        assertEq(borrowRate,    0.050000000000000000000050000e27);  // 1 / 499.999902325764663556e18 = 2e-21 => 2e-21/0.8 * 0.02 = 5e4 in ray
-        assertEq(liquidityRate, 0.000000000000000000000100000e27);  // 1 / 499.999902325764663556e18 = 2e-21 => 2e-21 * (0.05e27 + 5e4) = 1e5 in ray
+        assertEq(borrowRate,    0.050000000000000000000050000e27);  // 1 / 499.999902325764663556e18 = 2e-21 => 2e-21/0.8 * 0.02 = 5e4 in ray (rounding)
+        assertEq(liquidityRate, 0.000000000000000000000095000e27);  // 1 / 499.999902325764663556e18 = 2e-21 => 2e-21 * (0.05e27 + 5e4) = 1e5 in ray (rounding) TODO: Investigate why not 95%
```
Comment (barrutko) [inline test/RepayWithATokens.t.sol:679]: Is it not 95%?

File: test/RepayWithATokens.t.sol [hunk 7]
```diff
@@ -660,24 +662,28 @@ contract RepayWithATokensConcreteTests is RepayWithATokensTestBase {
...
 
         poolParams.liquidityIndex            = expectedLiquidityIndex;
         poolParams.variableBorrowIndex       = expectedVariableBorrowIndex;
         poolParams.currentLiquidityRate      = liquidityRate;
-        poolParams.currentVariableBorrowRate = borrowRate;
+        poolParams.currentVariableBorrowRate = borrowRate;
```
Comment (barrutko) [inline test/RepayWithATokens.t.sol:684]:
```suggestion
        poolParams.currentVariableBorrowRate = borrowRate;
```

File: test/Withdraw.t.sol [hunk 8]
```diff
@@ -298,17 +298,17 @@ contract WithdrawConcreteTests is WithdrawTestBase {
     {
         ( uint256 borrowRate, uint256 liquidityRate ) = _getUpdatedRates(100 ether, 1000 ether);
 
-        assertEq(borrowRate,    0.0525e27);   // 5% + 10%/80% of 2% = 5.25%
-        assertEq(liquidityRate, 0.00525e27);  // 10% of 5.25%
+        assertEq(borrowRate,    0.0525e27);          // 5% + 10%/80% of 2% = 5.25%
```
Comment (barrutko) [inline test/Withdraw.t.sol:301]: Too many spaces

**Review by barrutko — CHANGES_REQUESTED** [review]

File: test/Borrow.t.sol [hunk 9]
```diff
@@ -472,8 +472,8 @@ contract BorrowConcreteTests is BorrowTestBase {
         vm.prank(borrower);
         pool.borrow(address(borrowAsset), 500 ether, 2, 0, borrower);
 
-        poolParams.currentLiquidityRate      = 0.03125e27;  // 50% utilized: 6.25% * 50% = 3.125%
-        poolParams.currentVariableBorrowRate = 0.0625e27;   // 50% utilized: 5% + 50%/80% * 2% = 6.25%
+        poolParams.currentLiquidityRate      = 0.03125e27 * 0.95;  // 50% utilized: 6.25% * 50% = 3.125% * (1 - reserveFactor)
+        poolParams.currentVariableBorrowRate = 0.0625e27;          // 50% utilized: 5% + 50%/80% * 2% = 6.25%TIL
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply test/Borrow.t.sol:476]: Yeah cool right?

[the remark this answers is not in the document]
Comment (lucas-manuel) [reply test/Borrow.t.sol:476]: Removed

File: test/Borrow.t.sol [hunk 10]
```diff
@@ -472,8 +472,8 @@ contract BorrowConcreteTests is BorrowTestBase {
         vm.prank(borrower);
         pool.borrow(address(borrowAsset), 500 ether, 2, 0, borrower);
 
-        poolParams.currentLiquidityRate      = 0.03125e27;  // 50% utilized: 6.25% * 50% = 3.125%
-        poolParams.currentVariableBorrowRate = 0.0625e27;   // 50% utilized: 5% + 50%/80% * 2% = 6.25%
+        poolParams.currentLiquidityRate      = 0.03125e27 * 0.95;  // 50% utilized: 6.25% * 50% = 3.125% * (1 - reserveFactor)
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply test/Borrow.t.sol:475]: Yeah fair I'll make this change everywhere

File: test/FlashLoan.t.sol [hunk 11]
```diff
@@ -500,12 +500,12 @@ contract FlashLoanSuccessTests is FlashLoanTestBase {
 
         // Utilization rate has decreased because of premium being added to the pool
         assertLt(borrowRate,    0.0525e27);
-        assertLt(liquidityRate, 0.00525e27);
+        assertLt(liquidityRate, 0.0049875e27);   // 5.25% * 95%
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply test/FlashLoan.t.sol:503]: Good catch, fixed

File: test/FlashLoan.t.sol (hunk 1 above)
Comment (lucas-manuel) [reply test/FlashLoan.t.sol:885]: Fixed

File: test/FlashLoanSimple.t.sol (hunk 2 above)
Comment (lucas-manuel) [reply test/FlashLoanSimple.t.sol:414]: Think whitespace is gone now, forgot I turned off trim whitespace setting

File: test/FlashLoanSimple.t.sol (hunk 3 above)
Comment (lucas-manuel) [reply test/FlashLoanSimple.t.sol:484]: Fixed

File: test/RepayWithATokens.t.sol (hunk 6 above)
Comment (lucas-manuel) [reply test/RepayWithATokens.t.sol:679]: Yeah you're right removed

File: test/RepayWithATokens.t.sol (hunk 7 above)
Comment (lucas-manuel) [reply test/RepayWithATokens.t.sol:684]: Fixed

File: test/Withdraw.t.sol (hunk 8 above)
Comment (lucas-manuel) [reply test/Withdraw.t.sol:301]: Fixed

**Review by barrutko — APPROVED** [review]
Updated version looks good. I'll reapprove once up to date with master.

File: test/Borrow.t.sol [hunk 12]
```diff
@@ -472,8 +472,8 @@ contract BorrowConcreteTests is BorrowTestBase {
         vm.prank(borrower);
         pool.borrow(address(borrowAsset), 500 ether, 2, 0, borrower);
 
-        poolParams.currentLiquidityRate      = 0.03125e27;  // 50% utilized: 6.25% * 50% = 3.125%
-        poolParams.currentVariableBorrowRate = 0.0625e27;   // 50% utilized: 5% + 50%/80% * 2% = 6.25%
+        poolParams.currentLiquidityRate      = 0.03125e27 * 0.95;  // 50% utilized: 6.25% * 50% *  * (1 - reserveFactor) = 3.125% * 95%
```
Comment (hexonaut) [inline test/Borrow.t.sol:475]: Two multiplications in the comment. `50% *  * (1 - reserveFactor)` should be `50% * (1 - reserveFactor)`

**Review by hexonaut — DISMISSED** [review]
One minor thing then LGTM.

**Review by barrutko — DISMISSED** [review]

**Review by barrutko — APPROVED** [review]

**Review by hexonaut — APPROVED** [review]

**Pull request merged**